### PR TITLE
TASK-58023: Extract the anti-bruteforce base code into an addon

### PR DESCRIPTION
**Pull request by hakermi**
- Prepare anti-bruteforce addon base project: dependencies, modules and configs
- Extract the anti-bruteforce base code into an addon

### CODE REVIEW

File: anti-bruteforce-services/pom.xml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <artifactId>anti-bruteforce</artifactId>
+        <groupId>org.exoplatfrom</groupId>
+        <version>1.0.x-SNAPSHOT</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>anti-bruteforce-services</artifactId>
```
Comment (ahamdi) [inline anti-bruteforce-services/pom.xml:12]: Could you add the name tag here ?
```suggestion
    <artifactId>anti-bruteforce-services</artifactId>
    <name>eXo Anti-bruteforce - Services</name>
```

File: anti-bruteforce-webapp/pom.xml [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <artifactId>anti-bruteforce</artifactId>
+        <groupId>org.exoplatfrom</groupId>
+        <version>1.0.x-SNAPSHOT</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+    <packaging>war</packaging>
+
+    <artifactId>anti-bruteforce-webapp</artifactId>
```
Comment (ahamdi) [inline anti-bruteforce-webapp/pom.xml:13]:
```suggestion
    <artifactId>anti-bruteforce-webapp</artifactId>
    <name>eXo Anti-bruteforce - Webapp</name>
```

File: pom.xml [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    <parent>
+        <artifactId>addons-parent-pom</artifactId>
+        <groupId>org.exoplatform.addons</groupId>
+        <version>16-M01</version>
+    </parent>
+
+    <groupId>org.exoplatfrom</groupId>
+    <artifactId>anti-bruteforce</artifactId>
+    <packaging>pom</packaging>
+    <version>1.0.x-SNAPSHOT</version>
```
Comment (ahamdi) [inline pom.xml:15]:
```suggestion
    <version>1.0.x-SNAPSHOT</version>
    
    <name>eXo Anti-bruteforce - Parent</name>
```

**Review by ahamdi — COMMENTED** [review]
LGTM, except it lacks activation  configuration for the webapp config files.

**Review by rdenarie — COMMENTED** [review]
Hello,
Did you ensure that the authenticator plugin is correctly called in case of alternate authentication plugin ?
For example, if the metamask authentication plugin is active, do the antibruteforce plugin go correctly after the metamask authencation plugin ?

**Comment by hakermi** [comment]
> Hello, Did you ensure that the authenticator plugin is correctly called in case of alternate authentication plugin ? For example, if the metamask authentication plugin is active, do the antibruteforce plugin go correctly after the metamask authencation plugin ?

The metamask authenticater plugin doesn't use a username and a password, just validate the username (which is the wallet private key) to allow the login, so i don't see how the bruteforce attack will go here.
here we use an extendedautheticaterplugin, will work only if the user attempts to fill credentials (a password and a username , even those added when sign in with Metamask)
In fact, we didn't change anything even after the extracting, the code order execution will remain the same as before, you will see our changes on the organizationAuthenticaterIml soon.

File: anti-bruteforce-services/pom.xml (hunk 1 above)
Comment (hakermi) [reply anti-bruteforce-services/pom.xml:12]: done

**Comment by hakermi** [comment]
> LGTM, except it lacks activation configuration for the webapp config files.

You mean the portal config listener?, do we still need it even if we don't have pages ?

**Comment by boubaker** [comment]
> which is the wallet private key

@hakermi the privatekey isn't sent to server for the authentication, it's used by user to sign a message that will be sent instead. In fact, there is a big difference, because the private key of a wallet has to remain private and known by owner only, else the wallet will be compromised.

**Comment by hakermi** [comment]
> > which is the wallet private key
> 
> @hakermi the privatekey isn't sent to server for the authentication, it's used by user to sign a message that will be sent instead. In fact, there is a big difference, because the private key of a wallet has to remain private and known by owner only, else the wallet will be compromised.

OK, understood, Thanks for the explanation :),

**Comment by rdenarie** [comment]
For me, the way it is extracted is not the good one. 
We need to rediscuss.
Dont focus on metamask implem or what else. Imagine we can have a lot of authenticator plugin, each with a different behaviour. 
If all failed to authenticate the user, we need to count the failed attempts.
If the user failed attempts counter is greater than the max failed, we should block him BEFORE executing authenticator plugins.

Antibrute force IS NOT a new authenticator plugin

**Comment by hakermi** [comment]
> LGTM, except it lacks activation configuration for the webapp config files.

added

**Comment by ahamdi** [comment]
> For me, the way it is extracted is not the good one. We need to rediscuss. Dont focus on metamask implem or what else. Imagine we can have a lot of authenticator plugin, each with a different behaviour. If all failed to authenticate the user, we need to count the failed attempts. If the user failed attempts counter is greater than the max failed, we should block him BEFORE executing authenticator plugins.
> 
> Antibrute force IS NOT a new authenticator plugin

Indeed, we are adding a new plugin configuration which have the following role : 
 - Add some checks before attempting the login
 - Add a logic if a login is successfull
 - Add a logic if a login is failed
This plugin will help externalize work done for Anti-bruteforce addon and could be used to externalize disabled users feature.
Metamask does not pass by the usual login procedure and uses another plugin to validate. I think that we can not do a brute force attack using Metamask.
For all SSO providers, they do implement their anti-bruteforce mechanisms, thus I believe that our implementation should just focus on username/password authentication stored in eXo database.

**Comment by hakermi** [comment]
Since the name of the new plugin has created a confusion, we renamed it to a `SecurityCheckAuthenticationPlugin`. 
which is not exist yet, we will add it in the core module.

**Review by ahamdi — APPROVED** [review]

**Pull request merged**